### PR TITLE
Add Snowflake Client

### LINKED ISSUE #349: [FEATURE] Implement Snowflake client
**Is your feature request related to a problem? Please describe.**
We have a functional [Trino client](https://github.com/SneaksAndData/adapta/blob/main/adapta/storage/database/trino_sql.py), and I propose the addition of a Snowflake client offering similar capabilities. This new client could leverage the [Snowflake Python Connector](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-example) as a foundation.

**Describe the solution you'd like**
Starting with basic SQL support would suffice. Below is a reference example, using the TrinoClient.

```python
from adapta.storage.database.trino_sql import TrinoClient

os.environ['PROTEUS__TRINO_OAUTH2_USERNAME'] = 'user@mail.com'
client = TrinoClient(host="host.com", catalog="datalake")

query = """
SELECT 
    low_cardinality_feature
    , interesting_function(questionable_fact_column)
from cool_but_large_table
group by low_cardinality_feature
"""

with client:
    data = pd.concat(client.query(query))
```

### PR DESCRIPTION
**Pull request by adelinag08**
Closes #349 

## Scope

Implemented:
 - Adds a new `SnowflakeClient` class which provides a wrapper around the Snowflake Python connector. 
 - It contains a context manager to handle opening and closing the database connection
 - Includes a method for executing SQL queries and returning the result as a Pandas DataFrame.  

## Checklist

- [x] GitHub issue exists for this change.
- [x] Unit tests added and they pass.
- [x] Pylint 10.0/10.0 without bloating `.pylintrc` with exceptions.
- [x] Review requested on `latest` commit.

### CODE REVIEW

File: adapta/storage/database/snowflake_sql.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        :return: The SnowflakeClient instance, or None if there was an error connecting to the database.
+        """
+        try:
+            self._conn = snowflake.connector.connect(
+                user=self._user, account=self._account, warehouse=self._warehouse, authenticator=self._authenticator
+            )
+            return self
+        except DatabaseError as ex:
+            self._logger.error(
+                "Error connecting to {account} for {user}", host=self._account, user=self._user, exception=ex
```
Comment (george-zubrienko) [inline adapta/storage/database/snowflake_sql.py:59]:
```suggestion
                "Error connecting to {account} for {user}", account=self._account, user=self._user, exception=ex
```

File: adapta/storage/database/snowflake_sql.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        exc_tb: TracebackType | None,
+    ) -> None:
+        """
+        Exits the context manager and closes the database connection.
+
+        :param exc_type: The type of the exception that was raised, if any.
+        :param exc_val: The value of the exception that was raised, if any.
+        :param exc_tb: The traceback of the exception that was raised, if any.
+        """
+        self._conn.close()
```
Comment (george-zubrienko) [inline adapta/storage/database/snowflake_sql.py:76]: should we log the exc_val as well here?

File: adapta/storage/database/snowflake_sql.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        Executes the given SQL query and returns the result as a Pandas DataFrame.
+
+        :param query: The SQL query to execute.
+        :return: An iterator of Pandas DataFrames, one for each result set returned by the query, or None if there was
+            an error executing the query.
+        """
+        try:
+            with self._conn.cursor() as cursor:
+                cursor.execute(query)
+                return cursor.fetch_pandas_all()
```
Comment (george-zubrienko) [inline adapta/storage/database/snowflake_sql.py:89]:
```suggestion
                return cursor.execute(query).fetch_pandas_all()
```

File: pyproject.toml [hunk 4]
```diff
@@ -45,6 +45,8 @@ fsspec = { version = "2022.8.2" }
 
 trino = { version = "0.319.0", optional = true, extras = ['sqlalchemy'] }
 
+snowflake-connector-python = { version = "^3.4.0" }
```
Comment (george-zubrienko) [inline pyproject.toml:48]: you need to declare this as optional and add to the appropriate extras section. Remember to run `poetry update` after you apply the change

**Review by george-zubrienko — APPROVED** [review]

**Review by JanusAsmussen — APPROVED** [review]

**Pull request merged**